### PR TITLE
GPSW scheme

### PR DESCRIPTION
**Pull request by tilenmarc**
This pull request adds the first KP-ABE scheme based on [this paper](https://eprint.iacr.org/2006/309.pdf). The scheme uses pairing maps on elliptic curves, for which we use amcl library. To express the decoding policy a monotone span program (MSP) is needed, hence we implement a policy convertor which given a boolean expression describing the policy produces a MSP structure. Moreover, in the decryption process a matrix vector equation needs to be solved for which we implement a Gaussian elimination based process.

### CODE REVIEW

File: src/abe/policy.c [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+            cfe_mat_set(&inv_mat, one, i, i);
+        }
+
+        cfe_mat_copy(&msp_mat, msp->mat);
+        cfe_mat_mul(msp->mat, &msp_mat, &inv_mat);
+        cfe_mat_frees(&msp_mat, &inv_mat, NULL);
+    }
+    clearup:
+    cfe_vec_free(&vec);
+    mpz_clear(one);
```
Comment (janhartman) [inline src/abe/policy.c]: Free `bool_exp_trimmed`.

File: src/abe/policy.c [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+        cfe_mat_const(&inv_mat, msp->mat->cols, msp->mat->cols, zero);
+
+        for (size_t i = 0; i < inv_mat.rows; i++) {
+            cfe_mat_set(&inv_mat, one, 0, i);
+            cfe_mat_set(&inv_mat, one, i, i);
+        }
+
+        cfe_mat_copy(&msp_mat, msp->mat);
+        cfe_mat_mul(msp->mat, &msp_mat, &inv_mat);
+        cfe_mat_frees(&msp_mat, &inv_mat, NULL);
```
Comment (janhartman) [inline src/abe/policy.c]: `msp_mat` isn't needed, use `msp->mat` everywhere instead.

File: src/abe/policy.c [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    cfe_vec_const(vec2, c + 1, zero);
+    for (size_t i = 0; i < vec->size; i++) {
+        cfe_vec_set(vec2, vec->vec[i], i);
+    }
+    mpz_set_si(vec1->vec[c], -1);
+    mpz_set_si(vec2->vec[c], 1);
+    mpz_clear(zero);
+}
+
+int str_to_int(char *str) {
```
Comment (janhartman) [inline src/abe/policy.c]: Use https://en.cppreference.com/w/c/string/byte/strtol instead.

File: src/abe/policy.c [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+                cfe_mat_get(tmp, msp2.mat, i, j);
+                cfe_mat_set(msp->mat, tmp, i + msp1.mat->rows, j);
+            }
+            msp->row_to_attrib[i + msp1.mat->rows] = msp2.row_to_attrib[i];
+        }
+        return c_out;
+    }
+}
+
+void make_and_vecs(cfe_vec *vec1, cfe_vec *vec2, cfe_vec *vec, size_t c) {
```
Comment (janhartman) [inline src/abe/policy.c]: The name is not descriptive at all.... Rename or add a comment what it does (like in GoFE).

File: src/abe/policy.c [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+            continue;
+        }
+        if (bool_exp[i] == ')') {
+            num_brc--;
+            continue;
+        }
+        if (num_brc == 0 && i < strlen(bool_exp) - 3 && bool_exp[i] == 'A' &&
+                bool_exp[i+1] == 'N' && bool_exp[i+2] == 'D') {
+            bool_exp1 = substring(bool_exp, 0, i);
+            bool_exp2 = substring(bool_exp, i + 3, strlen(bool_exp));
```
Comment (janhartman) [inline src/abe/policy.c]: `bool_exp1` and `bool_exp2` are never freed.

File: src/abe/policy.c [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+            cfe_vec vec1, vec2;
+            make_and_vecs(&vec1, &vec2, vec, (size_t) c);
```
Comment (janhartman) [inline src/abe/policy.c]: `vec1 and `vec2` are never freed.

File: src/abe/policy.c [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+        for (size_t i = 0; i < msp2.mat->rows; i++) {
+            for (size_t j = 0; j < (size_t) c_out; j++) {
+                cfe_mat_get(tmp, msp2.mat, i, j);
+                cfe_mat_set(msp->mat, tmp, i + msp1.mat->rows, j);
+            }
+            msp->row_to_attrib[i + msp1.mat->rows] = msp2.row_to_attrib[i];
+        }
+        return c_out;
+    }
+}
```
Comment (janhartman) [inline src/abe/policy.c]: This function leaks memory everywhere. Make sure each memory initialization has a corresponding statement that frees it somewhere.

File: include/abe/policy.h [hunk 8]
```diff
@@ -50,7 +51,7 @@
  * vector [1, 1,..., 1] depending on the use case).
  */
 typedef struct cfe_msp {
-    cfe_mat *mat; /** A pointer to a matrix */
+    cfe_mat mat; /** A pointer to a matrix */
```
Comment (janhartman) [inline include/abe/policy.h]: Not a pointer any more :)

File: src/abe/gpsw.c [hunk 9]
```diff
@@ -138,29 +139,33 @@ void cfe_gpsw_delegate_keys(cfe_gpsw_keys *keys, cfe_vec_G1 *policy_keys,
...
     cfe_vec one_vec, alpha;
     mpz_t one;
     mpz_init_set_ui(one, 1);
-    cfe_vec_const(&one_vec, keys->mat.cols, one);
+    cfe_vec_init(&one_vec, keys->mat.cols);
+    cfe_vec_set_const(&one_vec, one);
     cfe_mat mat_transpose;
     cfe_mat_init(&mat_transpose, keys->mat.cols, keys->mat.rows);
     cfe_mat_transpose(&mat_transpose, &(keys->mat));
     int check = gaussian_elimination(&alpha, &mat_transpose, &one_vec, gpsw->p);
```
Comment (janhartman) [inline src/abe/gpsw.c]: `cfe_error check`

File: test/abe/policy.c [hunk 10]
```diff
@@ -31,42 +31,43 @@
 
 #include "sample/uniform.h"
 #include "abe/policy.h"
+#include <errno.h>
```
Comment (janhartman) [inline test/abe/policy.c]: Is this needed?

File: test/abe/policy.c [hunk 11]
```diff
@@ -101,16 +102,31 @@ MunitResult test_gaussian_elimination(const MunitParameter params[], void *data)
...
+    i = str_to_int(test_faulty_str);
+    munit_assert(i == -1);
+
+    return MUNIT_OK;
+}
+
 MunitTest policy_tests[] = {
         {(char *) "/test-boolean_to_msp",       test_boolean_to_msp,       NULL, NULL, MUNIT_TEST_OPTION_NONE, NULL},
-        {(char *) "/test-gaussian-elimination", test_gaussian_elimination, NULL, NULL, MUNIT_TEST_OPTION_NONE, NULL},
+//        {(char *) "/test-gaussian-elimination", test_gaussian_elimination, NULL, NULL, MUNIT_TEST_OPTION_NONE, NULL},
```
Comment (janhartman) [inline test/abe/policy.c]: Uncomment these tests.

**Comment by tilenmarc** [comment]
Thanks Jan, good catches and also thanks for the Valgrind reference. Now all the fixes should be done.

**Comment by janhartman** [comment]
Great!

**Comment by janhartman** [comment]
Don't forget to also push the updated documentation :)

**Pull request merged**